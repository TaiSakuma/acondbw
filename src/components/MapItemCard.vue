<template>
  <div class="mapitemcard">
    <div v-if="$apollo.queries.map.loading">loading...</div>
    <div v-else-if="error">Error: cannot load data</div>
    <v-card v-else-if="map" outlined hover style="max-width: 980px;">
      <div @click="$emit('expand')" style="cursor: default;">
        <v-container fluid class="pa-0">
          <v-row class="ma-0 px-0">
            <v-col cols="12" md="4" class="pa-2">
              <div class="caption grey--text">Name</div>
              <div class="font-weight-medium primary--text">
                <span @click.stop>
                  <router-link :to="'/maps/item/' + map.name" v-text="map.name"></router-link>
                </span>
              </div>
            </v-col>
            <v-col cols="6" md="4" class="pa-2">
              <div class="caption grey--text">Date posted</div>
              <div v-text="map.datePosted"></div>
            </v-col>
            <v-col cols="5" md="3" class="pa-2">
              <div class="caption grey--text">Mapper</div>
              <div v-text="map.mapper"></div>
            </v-col>
            <v-col cols="1" md="1" align-self="end" class="pa-2">
              <div v-if="collapsible">
                <v-row align="start" justify="end" class="px-1">
                  <v-tooltip bottom open-delay="800">
                    <template v-slot:activator="{ on }">
                      <v-btn
                        icon
                        @click.stop="
                      collapsed ? $emit('expand') : $emit('collapse')
                    "
                        v-on="on"
                      >
                        <v-icon>
                          {{
                          collapsed ? "mdi-chevron-down" : "mdi-chevron-up"
                          }}
                        </v-icon>
                      </v-btn>
                    </template>
                    <span>{{ collapsed ? "Expand" : "Collapse" }}</span>
                  </v-tooltip>
                </v-row>
              </div>
            </v-col>
          </v-row>
          <v-expand-transition>
            <v-row class="mx-0 mb-3 px-0 collapsible" v-show="!(collapsible && collapsed)">
              <v-col cols="12" md="8" offset-md="4" class="pa-2">
                <div class="caption grey--text">Paths</div>
                <ul v-if="map.mapFilePaths">
                  <li
                    v-for="(edgep, index) in map.mapFilePaths.edges"
                    :key="index"
                    v-text="edgep.node.path"
                  ></li>
                </ul>
              </v-col>
              <v-col cols="12" md="8" offset-md="4" class="pa-2">
                <div class="caption grey--text">Note</div>
                <div>
                  <ul v-if="map.note">
                    <li
                      v-for="(line, index) in map.note.split('\n')"
                      :key="index"
                    >{{ line.replace(/^- */, "") }}</li>
                  </ul>
                </div>
              </v-col>
              <v-col cols="12" md="8" offset-md="4" class="pa-2">
                <div class="caption grey--text">Beams</div>
                <ul v-if="map.beams">
                  <li v-for="(edgep, index) in map.beams.edges" :key="index">
                    <router-link :to="'/beams/item/' + edgep.node.name" v-text="edgep.node.name"></router-link>
                  </li>
                </ul>
              </v-col>
            </v-row>
          </v-expand-transition>
        </v-container>
      </div>
    </v-card>
    <div v-else>Nothing to show here.</div>
  </div>
</template>

<script>
import MAP from "@/graphql/Map.gql";

export default {
  name: "MapItemCard",
  props: {
    mapName: { default: null },
    collapsed: { default: false },
    collapsible: { default: false }
  },
  data() {
    return {
      map: null,
      error: null
    };
  },
  apollo: {
    map: {
      query: MAP,
      variables() {
        return {
          name: this.mapName
        };
      },
      result(result) {
        this.error = null;
        if (result.error) {
          this.error = true;
        }
      }
    }
  }
};
</script>
