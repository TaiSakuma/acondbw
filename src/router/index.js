import Vue from "vue";
import VueRouter from "vue-router";
import Home from "@/views/Home.vue";

Vue.use(VueRouter);

const routes = [
  {
    path: "/",
    name: "home",
    component: Home
  },
  {
    path: "/about",
    name: "about",
    component: () => import(/* webpackChunkName: "about" */ "@/views/About.vue")
  },
  {
    path: "/maps",
    component: () => import(/* webpackChunkName: "maps" */ "@/views/Maps.vue"),
    children: [
      {
        path: "",
        name: "MapList",
        component: () =>
          import(/* webpackChunkName: "maps" */ "@/views/MapList.vue")
      },
      {
        path: "item/:name",
        name: "MapItem",
        component: () =>
          import(/* webpackChunkName: "maps" */ "@/views/MapItem.vue")
      }
    ]
  },
  {
    path: "/beams",
    component: () => import(/* webpackChunkName: "beams" */ "@/views/Beams.vue"),
    children: [
      {
        path: "",
        name: "BeamList",
        component: () =>
          import(/* webpackChunkName: "beams" */ "@/views/BeamList.vue")
      },
      {
        path: "item/:name",
        name: "BeamItem",
        component: () =>
          import(/* webpackChunkName: "beams" */ "@/views/BeamItem.vue")
      }
    ]
  },
  {
    path: "/beams",
    name: "beam",
    component: () => import(/* webpackChunkName: "beams" */ "@/views/Beams.vue")
  }
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes
});

export default router;
