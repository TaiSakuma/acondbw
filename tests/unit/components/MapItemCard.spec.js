import Vue from "vue";
import VueRouter from "vue-router";
import Vuetify from "vuetify";
import { mount, createLocalVue } from "@vue/test-utils";

import MapItemCard from "@/components/MapItemCard.vue";
import router from "@/router";

Vue.use(Vuetify);
Vue.use(VueRouter);

describe("MapItemCard.vue", () => {
  let localVue;
  let vuetify;

  function createWrapper({ loading = false, propsData } = {}) {
    const mutate = jest.fn();
    let wrapper = mount(MapItemCard, {
      localVue,
      router,
      vuetify,
      mocks: {
        $apollo: {
          queries: {
            map: {
              loading: loading
            }
          },
          mutate
        }
      },
      propsData: {
        mapName: "lat20200201",
        ...propsData
      }
    });

    return wrapper;
  }

  const map = {
    mapId: "1013",
    name: "lat20200201",
    datePosted: "2020-02-01",
    mapper: "pwg-pmn",
    mapFilePaths: {
      edges: [
        {
          node: {
            path: "nersc:/go/to/my/maps_v3",
            note: "lat only"
          }
        }
      ]
    },
    note:
      "- This is a dummy test with a lat map                         \n- A beam depends on this map",
    beams: {
      edges: [
        {
          node: {
            beamId: "1150",
            name: "20200207"
          }
        }
      ]
    }
  };

  beforeEach(function() {
    localVue = createLocalVue();
    vuetify = new Vuetify();
  });

  it("loading", async () => {
    const loading = true;
    const wrapper = createWrapper({ loading });
    await Vue.nextTick();
    expect(wrapper.text()).toContain("loading");
  });

  it("error", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      error: true
    });
    await Vue.nextTick();
    expect(wrapper.text()).toContain("Error: cannot load data");
  });

  it("none", async () => {
    const wrapper = createWrapper();
    await Vue.nextTick();
    expect(wrapper.text()).toContain("Nothing to show here");
  });

  it("match snapshot", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      map: map
    });
    await Vue.nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it.each([
    [true, true, false],
    [true, false, true],
    [false, true, true],
    [false, false, true]
  ])(
    "collapsed - {collapsible: %p, collapsed: %p, visible: %p}",
    async (collapsible, collapsed, visible) => {
      const wrapper = createWrapper({
        propsData: { collapsed: collapsed, collapsible: collapsible }
      });
      wrapper.setData({
        map: map
      });
      await Vue.nextTick();
      expect(wrapper.find(".collapsible").isVisible()).toBe(visible);
    }
  );

  it("delete", async () => {
    const wrapper = createWrapper();
    wrapper.setData({
      map: map
    });
    await wrapper.vm.deleteMap();
    expect(wrapper.vm.$apollo.mutate).toBeCalled();
    expect(wrapper.emitted("mutated")).toBeTruthy();
  });
});
