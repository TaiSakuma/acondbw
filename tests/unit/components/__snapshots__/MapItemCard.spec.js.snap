// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapItemCard.vue match snapshot 1`] = `
<div class="mapitemcard" mapname="lat20200201">
  <div class="v-card v-card--hover v-card--outlined v-sheet theme--light" style="max-width: 980px;"><a href="/maps/item/lat20190213" class="">lat20190213</a> <a href="/maps/item/lat20200120" class="">lat20200120</a>
    <div class="" style="cursor: default;">
      <div class="container pa-0 container--fluid">
        <div class="row ma-0 px-0">
          <div class="py-0 col-md-4 col-10 order-1">
            <div class="caption grey--text">Name</div>
            <div class="font-weight-medium primary--text"><span><a href="/maps/item/lat20200201" class="">lat20200201</a></span></div>
          </div>
          <div class="py-0 col-md-4 col-6 order-3">
            <div class="caption grey--text">Date posted</div>
            <div>2020-02-01</div>
          </div>
          <div class="py-0 col-md-2 col-6 order-4">
            <div class="caption grey--text">Mapper</div>
            <div>pwg-pmn</div>
          </div>
          <div class="py-0 order-md-5 col col-2 order-2 align-self-end">
            <div class="row px-1 py-0 align-start justify-end">
              <!----> <span><button type="button" class="v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-dots-vertical theme--light"></i></span></button>
              <div class="v-menu">
                <!---->
              </div>
              </span></div>
          </div>
        </div>
        <div class="row mx-0 mb-3 px-0 collapsible">
          <div class="py-0 col-md-8 offset-md-4 col-12">
            <div class="caption grey--text">Paths</div>
            <ul>
              <li>nersc:/go/to/my/maps_v3</li>
            </ul>
          </div>
          <div class="py-0 col-md-8 offset-md-4 col-12">
            <div class="caption grey--text">Note</div>
            <div>
              <ul>
                <li>This is a dummy test with a lat map </li>
                <li>A beam depends on this map</li>
              </ul>
            </div>
          </div>
          <div class="py-0 col col-12">
            <div class="row">
              <div class="py-0 col-md-4 order-md-0 col-12 order-2 align-self-end"><span class="grey--text" style="font-size: 65%;">Data ID: </span></div>
              <div class="py-0 col-md-8 col-12 order-1">
                <div class="caption grey--text">Beams</div>
                <ul>
                  <li><a href="/beams/item/20200207" class="">20200207</a></li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
